crates/transaction-pool/src/pool/mod.rs

```diff
@@ -669,6 +671,25 @@ where
         removed
     }
 
+    /// Prune known senders which do not exist in any subpool
+    pub(crate) fn prune_sender_info(
+        &self,
+        outcome: &OnNewCanonicalStateOutcome<T::Transaction>
+    ) {
+        for sender_address in self.unique_senders().iter() {
+            for tx in self.get_transactions_by_sender(*sender_address) {
+                let sender_id = self.get_sender_id(*sender_address);
+                if outcome.mined.contains(tx.hash()) || outcome.discarded.iter().any(|d| d.sender_id() == sender_id) {
+                    let mut pool = self.pool.write();
+                    pool.remove_sender_info(&sender_id);
+                    let mut identifiers = self.identifiers.write();
+                    identifiers.remove_sender_address(&sender_id);
+                    identifiers.remove_sender_id(sender_address);
+                }
+            }
+        }
```

Comment:
we don't need the `OnNewCanonicalStateOutcome`

instead we can do this in two steps:

1. get all unique sender ids that currently have txs in the pool
2. retain only those in `self.identifiers`